File: java/programmers/src/괄호변환.java
import java.util.*;
public class 괄호변환 {
    StringBuilder sb = new StringBuilder();
    public String solution(String p) {

        if(checkRight(p)) return p;
        String answer = forU(p);
        return answer;
    }
    public String forU(String str){

        if(str.length()==0){
            return "";
        }
        int leftCnt =0;
        int rightCnt =0;
        String tmp ="";
        for(int i=0;i<str.length();i++){
            if(str.charAt(i)=='('){
                leftCnt++;
            }
            if(str.charAt(i)==')'){
                rightCnt++;
            }
            if(leftCnt==rightCnt){ //균형잡힌
                if(checkRight(str.substring(0,i+1))){ //올바르다면

                    tmp= str.substring(0,i+1)+forU(str.substring(i+1,str.length()));
                }
                else{

                    tmp= "("+forU(str.substring(i+1,str.length()))+")" + reverseStr(str.substring(1,i));
                }
                break;
            }
        }
        return tmp;
    }
    public String reverseStr(String str){

        String tmp  ="";
        for(int i=0;i<str.length();i++){
            if(str.charAt(i)=='('){
                tmp+=")";
            }
            else{
                tmp+="(";
            }
        }

        return tmp;
    }
    public boolean checkRight(String str){
        Stack<Character> st = new Stack<>();
        for(int i=0;i<str.length();i++){
            if(str.charAt(i)=='('){
                st.push(str.charAt(i));
            }
            if(str.charAt(i)==')') {
                if (st.isEmpty()) return false;
                st.pop();
            }
        }
        return true;
    }
}

File: c++/programmers/괄호변환.cpp
#include <string>
#include <vector>
#include <iostream>
#include <stack>
using namespace std;
int n;
string answer;
stack<int> st;
string lf;
string rt;
string start(int a)
{
    lf = "(";
    rt = ")";
    cout <<a <<"여까지옴n";
    if (a == n)
    {
        return "";
    }
    int l = 0;
    int r = 0;
    string tmp;
    while (true)
    {
        if (answer[a] == '(')
        {
            l++;
            tmp.push_back('(');
            st.push(0);
        }
        else
        {
            r++;
            tmp.push_back(')');
            if (!st.empty())
            {
                if (st.top() == 0)
                {
                    st.pop();
                }
                else
                {
                    st.push(1);
                }
            }
            else
            {
                st.push(1);
            }
        }
        if (l > 0 && r > 0)
        {
            if (r == l)
            {
                if (st.empty())
                { //올바른 괄호 문자열이라면
                    tmp += start(a + 1);
                }
                else // 균형잡힌 괄호 문자열이라면
                {
                    string tmp2 ="";
                    for(int i=1;i<tmp.size()-1;i++){
                        if(tmp[i] =='('){
                            tmp2.push_back(')');
                        }
                        else{
                            tmp2.push_back('(');
                        }

                    }
                    
                    return lf+start(a+1)+rt+tmp2;
                }
                break;
            }
        }
        a++;
    }
    return tmp;
}
string solution(string p)
{
    answer = "";

    n = p.size();
    if (n == 0)
    {
        return answer;
    }
    answer = start(0);

    return answer;
}
